### PR TITLE
Make LXC exec driver compatible with recent LXC versions

### PR DESCRIPTION
**Pull request by samm-git**
Fix for the https://github.com/docker/docker/issues/12912

Signed-off-by: Alex Samorukov samm@os2.kiev.ua

### CODE REVIEW

**Comment by jfrazelle** [comment]
ping @ashahab-altiscale

**Comment by ashahab-altiscale** [comment]
@samm-git Thanks a lot for the PR! What kind of machines and kernels have you tested this in?

**Comment by ashahab-altiscale** [comment]
Also can you update the Dockerfile so that it compiles with LXC-1.1.2?

**Comment by samm-git** [comment]
@ashahab-altiscale I was testing on OpenWRT/PPC, but I believe that problem is non OS/Kernel depended - LXC 1.1.0 and later has lxc.autodev on by default and it breaks mount in lxc-start. 

About updating Dockerfile - I will do and test.

**Comment by samm-git** [comment]
@ashahab-altiscale i tested and confirmed bug on Ubuntu 14.04.2 LTS with docker 1.6.0 (from docker repository) and LXC 1.1.2 (self compiled). Result provided below:
```
root@IL1P1MGDOCKER02:~# docker run -ti busybox sh
lxc-start: conf.c: mount_entry: 1722 No such file or directory - failed to mount '/dev/pts/1' on '/usr/local/lib/lxc/rootfs//dev/console'
lxc-start: conf.c: lxc_setup: 3843 failed to setup the mount entries for '5fac669f7caf532820f8acbe9646068fe824ac93bb5e8d6e772d6ca599c48523'
lxc-start: start.c: do_start: 699 failed to setup the container
lxc-start: sync.c: __sync_wait: 51 invalid sequence number 1. expected 2
lxc-start: start.c: __lxc_start: 1164 failed to spawn '5fac669f7caf532820f8acbe9646068fe824ac93bb5e8d6e772d6ca599c48523'
lxc-start: lxc_start.c: main: 344 The container failed to start.
lxc-start: lxc_start.c: main: 348 Additional information can be obtained by setting the --logfile and --logpriority options.
```

With patched docker from trunk (1.7-dev) everything works as it should:
```
root@IL1P1MGDOCKER02:~# docker run -ti busybox sh
/ #
```
Also i added LXC 1.1.2 to the Dockerfile as you requested

**Comment by runcom** [comment]
Gorgon went nuts

**Comment by ashahab-altiscale** [comment]
@samm-git Looks good. Containers came up for me, now I'm running tests. I like the patch so far.
Ping @crosbymichael

**Comment by ashahab-altiscale** [comment]
@crosbymichael Can this be merged? It's passing all tests and is going to allow us to use lxc-1.1.2

**Comment by jfrazelle** [comment]
we need to test on jenkins

**Comment by jfrazelle** [comment]
then we need to update the docs to show we only use this version now in
packagers i think

**Comment by jfrazelle** [comment]
commits need DCO

**Comment by jfrazelle** [comment]
and squashing

**Comment by samm-git** [comment]
@jfrazelle could you please tell me what is the best way to do this without opening new PR? If opening new PR is a best way to do this please let me know.

**Comment by jfrazelle** [comment]
no please do not open a new PR

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

~~~console
$ git clone -b "master" git@github.com:samm-git/docker.git somewhere
$ cd somewhere
$ git remote add upstream git@github.com:docker/docker
$ git fetch master
$ git rebase -i upstream/master
# follow instructions in window to squash to one commit
$ git commit --amend -s --no-edit
$ git push -f
~~~
This will update the existing PR, so you do not need to open a new one.

**Comment by samm-git** [comment]
@jfrazelle finally done, thank you.

**Comment by cpuguy83** [comment]
looks like 1 lxc failure.

**Comment by samm-git** [comment]
@cpuguy83  any idea how to debug/fix this? i am using docker with lxc 1.1.2 w/o any issues in prod.

**Comment by cpuguy83** [comment]
@samm-git check with the other deivces tests and see if they are skipped on LXC.
It doesn't look like the LXC driver is creating the container path necessary for use with devices.

**Comment by samm-git** [comment]
@cpuguy83 thank you. I just tested - device is not created, but it is also missing in the config.lxc generated for the container, so in fact it is a Docker issue. It was working before because lxc.autodev was turned off (pre-1.1.0), but now, when its on docker need to put all such mounts in the lxc template.

**Comment by ashahab-altiscale** [comment]
@samm-git Yes, passing the list of devices to the template or to a hook seems to a good idea. Did you try running the test with autodev off like before?

**Comment by samm-git** [comment]
@ashahab-altiscale  it should work w/o autodev because docker will do the job. To make it work with autodev - script for the autodev hook needs to be created with appropriate mknod calls.

**Comment by samm-git** [comment]
@cpuguy83  test should pass now

**Comment by ashahab-altiscale** [comment]
@samm-git FAIL: docker_cli_run_test.go:3089: DockerSuite.TestRunReadProcLatency This is still failing.

**Comment by tiborvass** [comment]
Ping @samm-git

**Comment by samm-git** [comment]
@tiborvass  reproducing issue right now

**Comment by samm-git** [comment]
@ashahab-altiscale  as far as i could see - this is expected:
```go
func (s *DockerSuite) TestRunReadProcLatency(c *check.C) {
	testRequires(c, NativeExecDriver)
```

**Comment by samm-git** [comment]
@tiborvass  @ashahab-altiscale  i analysed test failures and i dont think that they were related to my changes. Also failing tests are skipped in the trunk. Could you please run this VM again?

**Comment by ashahab-altiscale** [comment]
@samm-git Thanks.

**Comment by ashahab-altiscale** [comment]
@jfrazelle Why is the TestRunReadProcLatency is running in lxc jenkins.

**Comment by samm-git** [comment]
@ashahab-altiscale @jfrazelle  ping

**Comment by ashahab-altiscale** [comment]
@samm-git @jfrazelle  If the tests are skipped in trunk, why aren't they being skipped here?

**Comment by jfrazelle** [comment]
I am rerunning the lxc tesst now

**Comment by jfrazelle** [comment]
passed!

File: daemon/execdriver/lxc/lxc_template.go [hunk 1]
```diff
@@ -61,6 +61,9 @@ lxc.cgroup.devices.allow = {{$allowedDevice.CgroupString}}
 # Use mnt.putold as per https://bugs.launchpad.net/ubuntu/+source/lxc/+bug/986385
 lxc.pivotdir = lxc_putold
 
+# lxc.autodev is not compativle with lxc --device switch
```
Comment (LK4D4) [inline daemon/execdriver/lxc/lxc_template.go]: combatible

**Comment by LK4D4** [comment]
Tests passed
LGTM apart from misspell

**Comment by icecrime** [comment]
Thanks @samm-git! Carried in #13976 (I preserved your commit).

**Pull request closed without merging**